### PR TITLE
Update sort functions

### LINKED ISSUE #1945: Update sort functions for `List` and `Array`
Currently both `List` and `Array` only have a `sortWith` function which returns `Int32` (this should be changed to `Comparison`).
We discussed that we should add a function `sort : List[a] -> List[a] with Order[a]` (the same for `Array`).

In addition, we probably want to add the following [`sortOn`](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-List.html#v:sortOn) function from Haskell for `List` (and `Array`):
```
sortOn :: Ord b => (a -> b) -> [a] -> [a] 
```
This can for example be used to sort on a specific field of a record:
```
let l = {i=2, s="World"}::{i=1, s="Hello"}::Nil;
let l' = sortOn (x -> x.i) l;
l' == {i=1, s="Hello"}::{i=2, s="World"}::Nil
```
There are currently no "sort" functions outside `List` and `Array`.

Any comments @magnus-madsen ?

### PR DESCRIPTION
**Pull request by felixwiemuth**
Resolves #1945

### CODE REVIEW

**Review by felixwiemuth — COMMENTED** [review]
The `sortWith` tests have to be updated because of the change of the comparison function type. In addition, tests for all sort functions should be added.

But I suggest that the changes are reviewed first before I update/add tests.

**Review by magnus-madsen — COMMENTED** [review]
Structure looks good

File: main/src/library/Array.flix [hunk 1]
```diff
@@ -1584,13 +1584,26 @@ namespace Array {
         foreachWithIndex(step, a);
         StringBuilder.toString(sb)
 
+    ///
+    /// Sort array `a` using the `Order` instance on the elements.
```
[the remark this answers is not in the document]
Comment (felixwiemuth) [reply main/src/library/Array.flix:1588]: Good point.

[the remark this answers is not in the document]
Comment (felixwiemuth) [reply main/src/library/Array.flix:1588]: I actually just saw that there is also a `sortWith!` which sorts in-place, so it makes sense to have the same variants `sort!` and `sortBy!` as for `sort`.

**Comment by felixwiemuth** [comment]
This is replaced by #1997 (the issue number in the branch name was also wrong here).

**Pull request closed without merging**